dev/ColorPicker/readme.md

```diff
@@ -0,0 +1,141 @@
+
+1. `ColorPicker` : The top-level control containing all other controls. This is the control intended for user interaction and used by application developers.
+2. `ColorPickerSlider` : A custom slider implementation used for the sliders in the `ColorPicker` control template itself. These sliders are unique in that they have a gradient background rendered to correspond with a defined HSV color channel. Sliders are 1-dimensional. This control is not useful apart from the `ColorPicker`.
+3. `ColorSpectrum` : The main graphic displaying a field of possible colors the user can select from. The color spectrum is 2-dimensional and can be viewed in a Box or Ring shape.
+4. `SpectrumBrush` : A special brush used to render the color spectrum background using composition APIs. This is only available on Windows version 1703 (RS2) and later.
+
+The control hierarchy and namespaces are as follows:
+
+```
+ ─ ColorPicker             [Microsoft.UI.Xaml.Controls]
+   ├── ColorPickerSlider   [Microsoft.UI.Xaml.Controls.Primitives]
+   └── ColorSpectrum       [Microsoft.UI.Xaml.Controls.Primitives]
+       └── SpectrumBrush   [Microsoft.UI.Xaml.Controls.Primitives]
+```
+
+Generalized helper methods are also located in two places:
+
+ * ColorHelpers.[cpp/h] : Helpers specifically for the color picker and related controls. These are located in the `dev/ColorPicker` directory itself.
+ * ColorConversion.[cpp/h] : More generalized helpers for working with RGB and HSV colors. The RGB and HSV structures are defined here as well as conversions between the two color representations and hex. These files are located in the `dev/Common` directory separate from ColorPicker.
+
+## ColorPicker
+
+[Control Properties](https://docs.microsoft.com/en-us/uwp/api/microsoft.ui.xaml.controls.colorpicker#properties), [Control Events](https://docs.microsoft.com/en-us/uwp/api/microsoft.ui.xaml.controls.colorpicker#events)
+
+Most of the detailed functionality specific to the `ColorPicker` is implemented in the primitive controls outlined below. See the links above and at the top of this document for full details on the usage of the `ColorPicker` itself.
+
+The `ColorPicker` functions by monitoring for changing in properties or user selection and then updating the selected color and graphical controls (preview rectangles, sliders, input text boxes, etc.) accordingly. It also ensures the various properties and input is validated before computing the updated selected color.
+
+Internally, the `ColorPicker` always represents the selected color in HSV instead of RGB. For cases where precision is important this creates some problems as the color is always converted to RGB before being available from external APIs. This is in contrast to the `ColorPickerSlider` and `ColorSpectrum` which pass the color information directly in HSV to avoid loss of precision. 
```

Comment:
It doesn't harm anything to have this available externally. If there was a scenario where it's possible to edit the same selected color using two different pickers this would be important. Further value would be in adding an HsvColor struct as well. Fundamentally, I don't think WinUI should make an assumption of how the control is used by applications and artificially restrict accuracy like this.

Reply:
I definitely agree with you in concept and exposing an HsvColor with a Vector4 type is not a great API. However, there are lots of other examples of this including in the Community Toolkit. I think there is enough to design something clean and future-proof here. Regardless, I've already broken this out into a feature request #3281 as it isn't a lesson's learned or something to keep in mind doing a clean re-implementation. 

This is actually a good time to advocate again for keeping the lesson's learned in this document. They are changes that should be made to make the design cleaner -- however, changing that probably won't happen as they would be breaking. It is definitely something for posterity to consider in other implementations though. If the WPF team left notes like that UWP would have been a lot better I'm sure.

Reply:
How about adding lessons learned in an improvements.md file? That way everyone knows where to look for those kind of things. And whether that is in readme.md or another file does not make a difference regarding whether one would actually learn from that. If WPF had that in readme.md files or in improvements.md files wouldn't have made a difference I think. What do you think about that @robloo ?

Reply:
I have no concern moving to a separate file. The information would still be retained alongside the code. I can justify creating a separate file as well since the topic is somewhat different from architecture. Should this be in a Docs folder though? Not sure the criteria decided for NavView.

Reply:
In the case of NavView, I created a docs folder since we also needed to document NavViewItem and rendering wasn't easy to explain. I think in the case of ColorPicker, just putting a separate file at the same level of the readme is probably fine.